html/dom/dynamic-markup-insertion/opening-the-input-stream/event-listeners.window.js

```diff
@@ -0,0 +1,42 @@
+test(t => {
+  const frame = document.body.appendChild(document.createElement("iframe")),
+        body = frame.contentDocument.body;
+  t.add_cleanup(() => frame.remove());
+  frame.contentDocument.addEventListener("x", t.unreached_func());
+  body.addEventListener("x", t.unreached_func());
+  frame.contentDocument.open();
+  frame.contentDocument.close();
+  frame.contentDocument.dispatchEvent(new Event("x"));
+  body.dispatchEvent(new Event("x"));
+}, "Event listeners are to be removed");
+
+test(t => {
+  const frame = document.body.appendChild(document.createElement("iframe"));
+  t.add_cleanup(() => frame.remove());
+  let once = false;
+  frame.contentDocument.addEventListener("x", () => {
+    frame.contentDocument.open();
+    frame.contentDocument.close();
+    once = true;
+  });
+  frame.contentDocument.addEventListener("x", t.unreached_func());
+  frame.contentDocument.dispatchEvent(new Event("x"));
+  assert_true(once);
+}, "Event listeners are to be removed with immediate effect");
+
+test(t => {
```

Comment:
Might be worth gating this behind an `if (document.body.attachShadow)` so browsers without shadow DOM don't get failures.

Reply:
I think that's bad practice. It makes it less clear what's going on. And have a varying number of tests between browsers, if that's what you're suggesting, seems even worse for tooling.